packages/integration-sdk-runtime/src/synchronization/batchBySize.ts

```diff
@@ -0,0 +1,117 @@
+  if (getSizeOfObject(data) < sizeInBytes) {
+    return [data]; //Just one chunk of all data
+  }
+  let bestIndex = binarySearch(data, sizeInBytes);
+  if (bestIndex <= 0) {
+    //If the first entity is too big
+    handleBinarySearchError(data, sizeInBytes, logger); // we remove rawdata
+    bestIndex = 1; //and send it to the persister alone. TODO: find a way to avoid sending this entities as a single call. It shouldn't be an issue since this only happens if there is a single entity above sizeInBytes
+  }
+  chunkedData.push(data.slice(0, bestIndex));
+  if (bestIndex !== data.length)
+    chunkedData.push(
+      ...chunk(data.slice(bestIndex), sizeInBytes, logger), //chunk the rest of the data
+    );
+  return chunkedData;
+}
+//If the batch doesnt fit, we try with half, and then half of half ... until it fits.
+//We repeat until we reach the Threshold or until the while condition.
+//We are binary-searching for the best spot to chunk
+function binarySearch<T extends UploadDataLookup, K extends keyof T>(
+  data: T[K][],
+  targetSize: number,
+): number {
+  let left: number = 0;
+  let right: number = data.length - 1;
+  let bestIndex: number = -1;
+  const thSize = targetSize * BATCH_THRESHOLD;
+  let mid: number = Math.floor((left + right) / 2);
+  let size = getSizeOfObject(data.slice(0, mid));
+  while (left <= right) {
+    if (size <= targetSize) {
+      bestIndex = mid;
+      if (size >= thSize) {
+        return mid; //if its **close enough** to the target size
+      }
+    }
+    if (targetSize < size) {
+      right = mid - 1;
+      if (left <= right) {
+        const newMid = Math.floor((left + right) / 2);
+        size = size - getSizeOfObject(data.slice(newMid, mid));
+        mid = newMid;
+      }
+    } else {
+      left = mid + 1;
+      if (left <= right) {
+        const newMid = Math.floor((left + right) / 2);
+        size = size + getSizeOfObject(data.slice(mid, newMid));
+        mid = newMid;
+      }
+    }
+  }
+  return bestIndex;
+}
+function handleBinarySearchError<T extends UploadDataLookup, K extends keyof T>(
+  data: T[K][],
+  sizeInBytes: number,
+  logger: IntegrationLogger,
+) {
+  const bigBatch = data.slice(0, 1) as Entity[] | Relationship[];
+  if (bigBatch[0] && bigBatch[0]._rawData) {
+    logger.info(
+      bigBatch[0].key,
+      'Entity/Relationship is too big to ingest. Will remove _rawData',
+    );
+    bigBatch[0]._rawData = [];
+  }
+  if (getSizeOfObject(bigBatch) > sizeInBytes) {
+    logger.error(
+      bigBatch[0].key,
+      'Entity/Relationship is too big to ingest. This usually happens when item has a lot of long attributes.',
+    );
+    throw new Error(
+      'Entity/Relationship is too big to ingest. This usually happens when item has a lot of long attributes.',
+    );
+  }
+}
+
+export function getSizeOfObject<T extends UploadDataLookup, K extends keyof T>(
+  object: T[K][],
+): number {
+  try {
+    return JSON.stringify(object).length;
```

Comment:
I'm not exactly sure, but I'm seeing some evidence from this project (https://github.com/shelfio/array-chunk-by-size/blob/master/src/index.ts#L62) that the size in bytes for an object may not exactly equal its string length:

```ts
function getObjectSize<T>(obj: T): number {
  try {
    const str = stringify(obj);

    return Buffer.byteLength(str, 'utf8');
  ...
}
```

Can you see if `.length` is equivalent to `Buffer.byteLength()`? Or if it's safer to use `Buffer.byteLength()`?

Reply:
I also notice that the same package I linked uses `json-stringify-safe` package, which is "Like JSON.stringify, but doesn't throw on circular references."

I don't expect our entities/relationships to have circular references, but it is probably worth doing the same here for safety.

Reply:
console.log("🌏".length)//2
console.log(Buffer.byteLength("🌏"))//4
console.log("a".length)//1
console.log(Buffer.byteLength("a"))//1